src/shell.ts

```diff
@@ -0,0 +1,170 @@
+import { T7Map } from './map'
+import { Terminal } from '@tuzig/xterm'
+import { loadCommands, Command } from './commands'
+import { Fields, Form } from './form'
+import { Clipboard } from "@capacitor/clipboard"
+
+export class Shell {
+
+    prompt = "T7> "
+
+    map: T7Map
+    t: Terminal
+    onKey: (ev: KeyboardEvent) => void
+    active = false
+    activeForm: Form | null
+    commands: Map<string, Command>
+    field = ''
+
+    constructor(map: T7Map) {
+        this.map = map
+        this.t = map.t0
+    }
+
+    start() {
+        if (this.active)
+            return
+        this.active = true
+        this.commands = loadCommands(this)
+        let field = ''
+        this.t.scrollToBottom()
+        this.onKey = async ev => {
+            const key = ev.key
+            switch (key) {
+                case "Enter":
+                    this.t.write("\n")
+                    await this.handleLine(field)
+                    field = ''
+                    break
+                case "Backspace":
+                    if (field.length > 0) {
+                        field = field.slice(0, -1)
+                        this.t.write("\b \b")
+                    }
+                    break
+                default:
+                    if (key.length == 1) { // make sure the key is a char
+                        field += key
+                        this.t.write(key)
+                    }
+            }
+            this.field = field
+        }
+    }
+
+    async handleLine(input: string) {
+        const [cmd, ...args] = input.trim().split(/\s+/)
+        await this.execute(cmd, args)
+        this.field = ''
+        this.printPrompt()
+    }
+
+    async execute(cmd: string, args: string[]) {
+        if (this.activeForm) 
+            this.escapeActiveForm()
+        if (!cmd)
+            return
+        const command = this.commands.get(cmd)
+        if (!command)
+            return this.t.writeln(`Command not found: ${cmd}`)
+        this.active = false
```

Comment:
So `active` really means `commandActive`? If it is, could be better to have an `activeCommand` property to point to the active command or null (just like activeForm)

Reply:
No - `active` means the shell is active, as in able to get input (i.e. no command *or* form active). Maybe we could have `activeCommand` instead but we will have to check for both `activeCommand` and `activeForm` when getting input (sometimes commands aren't finished but don't have a form active, e.g. `addHost`).